keps/sig-node/3721-support-for-env-files/README.md

```diff
@@ -0,0 +1,535 @@
+    volumeMounts:
+    - name: config
+      mountPath: /data
+   containers:
+  - name: use-envfile
+    image: registry.k8s.io/busybox
+    command: [ "/bin/sh", "-c", "env" ]
+    env:
+    - name: CONFIG_VAR
+      valueFrom:
+        fileKeyRef:
+          path: config.env
+          volumeName: config
+          key: CONFIG_VAR
+  restartPolicy: Never
+  volumes:
+  - name: config
+    emptyDir: {}
+```
+
+#### Story 2
+
+The user's workload needs to consume secrets of a service. Usually, they can
+store it in k8s Secret resource and use it to instantiate the env var. But it
+can become difficult to manage if each Pod needs to consume unique secrets.
+
+Instead, k8s can provide a new mechanism to fetch secrets in initContainer and
+store it in a file. The container can then use the `fileKeyRef` field to
+instantiate the env var from the file. Behind the scene, kubelet will parse
+the env file and instantiate these env vars during container creation. It
+does not mount this file onto the main container. In the example below,
+the Pod's output includes `JWT=token`.
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: dapi-test-pod
+spec:
+  initContainers:
+  - name: setup-envfile
+    image: registry.k8s.io/busybox
+    command: ['sh', '-c', 'echo "JWT=token" > /data/config.env']
+    volumeMounts:
+    - name: config
+      mountPath: /data
+   containers:
+  - name: use-envfile
+    image: registry.k8s.io/busybox
+    command: [ "/bin/sh", "-c", "env" ]
+    env:
+    - name: CONFIG_VAR
+      valueFrom:
+        fileKeyRef:
+          path: config.env
+          volumeName: config
+          key: JWT
+  restartPolicy: Never
+  volumes:
+  - name: config
+    emptyDir: {}
+```
+
+
+
+## Design Details
+
+Pods and PodTemplate will include `FileKeySelector` field under [`EnvVarSource`](https://pkg.go.dev/k8s.io/api/core/v1#EnvVarSource)
+that can be set for an individual container.
+
+### Pod API
+
+`FileKeySelector` field will allow the users to select the value for environment
+variable from the specified key in the file. The user can specify a file using `VolumeName` and
+`Path` fields. The `VolumeName` field specifies the volume mount that contains
+the file and `Path` is the relative path in this volume mount filesystem.
+
+
+```
+type EnvVarSource struct {
+    ...
+    // Selects a key of the env file.
+    // +optional
+    FileKeyRef *FileKeySelector `json:"fileKeyRef,omitempty" protobuf:"bytes,5,opt,name=fileKeyRef"`
+    ...
+}
+
+type FileKeySelector struct {
+    // The name of the volume mount containing the env file.
+    VolumeName string `json:",inline" protobuf:"bytes,1,opt,name=volumeName"`
+    // The relative file path inside the volume mount to select from.
+    Path string `json:",inline" protobuf:"bytes,2,opt,name=path"`
+    // The key of the env file to select from.  Must be a valid key.
+    Key string `json:"key" protobuf:"bytes,3,opt,name=key"`
```

Comment:
What happens if the specified key is not found at runtime?

What if the key is found but has no value?

Reply:
The general error handling should be discenible from the API docs